cranelift-codegen/src/redundant_reload_remover.rs

```diff
@@ -0,0 +1,462 @@
+// and destination are registers.  But that didn't cause any more fills to be removed or converted
+// to copies.  It's not clear why.
+//
+// It's also possible for a spill to be detected as redundant, if such a binding already appears
+// in the environment.  This was also implemented, but it was almost entirely ineffective, so was
+// removed again.
+
+// This describes a spill slot in the obvious way.  It also tracks the SSA value name associated
+// with the slot.
+#[derive(Clone, Copy)]
+struct SlotInfo {
+    value: Value,
+    offset: i32,
+    size: u32,
+}
+
+// Find out if the range `[offset, +size)` overlaps with the range in `si`.
+fn any_overlap(si: &SlotInfo, offset: i32, size: u32) -> bool {
+    let a_offset = si.offset as i64;
+    let a_size = si.size as i64;
+    let b_offset = offset as i64;
+    let b_size = size as i64;
+    let no_overlap = a_offset + a_size <= b_offset || b_offset + b_size <= a_offset;
+    !no_overlap
+}
+
+// Find, in `reginfo`, the register bank that `reg` lives in, and return the lower limit and size
+// of the bank.  This is so the caller can conveniently iterate over all `RegUnits` in the bank
+// that `reg` lives in.
+fn find_bank_limits(reginfo: &RegInfo, reg: RegUnit) -> (RegUnit, u16) {
+    if let Some(bank) = reginfo.bank_containing_regunit(reg) {
+        return (bank.first_unit, bank.units);
+    }
+    // We should never get here, since `reg` must come from *some* RegBank.
+    panic!("find_regclass_limits: reg not found");
+}
+
+// The redundant-fill remover processes each EBB independently.  It makes a single forward pass
+// over each EBB, carrying along a running environment, `AvailEnv`, that maps `RegUnits` to an
+// `Option<SlotInfo>`.
+//
+// If the triplet `(regunit, offset, size)` appears in the environment, this indicates that, at
+// this point, `regunit` and the stack slot denoted by `(offset, size)` have the same value.  We
+// also keep track of the SSA `Value` associated with the `reg`, because that is necessary to
+// create `copy` instructions that read `reg`.
+//
+// Invariants for the environment:
+//
+// The environment may have multiple different registers bound to the same stack slot -- that is,
+// `(offset, size)` pair.  That's OK, and reflects the reality that those two registers contain
+// the same value.
+//
+// The `value` fields must however be unique: it isn't allowable to have two entries in the
+// environment that have the same `value` field.  To do so would be to imply that that SSA value
+// has two different locations, since each environment entry is associated with a different
+// register, and so this would imply that the Value resided in both of those registers.  Function
+// `invariants_ok` checks this.
+//
+// It is also the case that the set of `(offset, size)` segments in the environment must be
+// non-overlapping.  `invariants_ok` doesn't check that, though.
+
+struct AvailEnv {
+    map: Vec<Option<SlotInfo>>,
+    size: usize,
```

Comment:
What's the point of storing the size, since it shouldn't change over time? To avoid calls to the `len` function of the map attribute? (it implies manual syncing between the actual size and the stored size attribute, which tends to make me a bit nervous)